src/librustc_typeck/check/expr.rs

```diff
@@ -1405,6 +1402,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 _ => {}
             }
 
+            let deref_t = match expr_t.kind {
+                ty::Ref(_, ref_t, _) => ref_t,
+                _ => &expr_t
+            };
+            match deref_t.kind {
+                ty::Adt(def, _) if !def.is_enum() => {
+                    self.suggest_fields_on_recordish(&mut err, def, field);
+                }
+                ty::Param(param_ty) => {
+                    self.explain_param(&mut err, param_ty);
+                }
+                _ => {}
+            }
+
             if field.name == kw::Await {
```

Comment:
Btw, I added this bit of code and I think it should perhaps be made into its own method as well... if you want to, that could be some nice drive-by cleanup. :)